components/server/src/raft_engine_switch.rs

```diff
@@ -1,126 +1,39 @@
-    flag_path
-}
-
-fn remove_tmp_dir<P: AsRef<Path>>(dir: P) {
-    match delete_dir_if_exist(&dir) {
-        Err(e) => warn!("Cannot remove {:?}: {}", dir.as_ref(), e),
-        Ok(true) => info!("Remove {:?} success", dir.as_ref()),
-        Ok(_) => {}
-    }
-}
-
-fn rename_to_tmp_dir<P1: AsRef<Path>, P2: AsRef<Path>>(src: P1, dst: P2) {
-    fs::rename(&src, &dst).unwrap();
-
-    let mut dir = dst.as_ref().to_path_buf();
-    assert!(dir.pop());
-    fs::File::open(&dir).and_then(|d| d.sync_all()).unwrap();
-    info!("Rename {:?} to {:?} correctly", src.as_ref(), dst.as_ref());
-}
-
-fn clear_raft_engine(engine: &RaftLogEngine) -> Result<(), EngineError> {
-    let mut batch_to_clean = engine.log_batch(0);
-    for id in engine.raft_groups() {
-        let state = engine.get_raft_state(id)?.unwrap();
-        engine.clean(id, 0, &state, &mut batch_to_clean)?;
-    }
-    engine.consume(&mut batch_to_clean, true).map(|_| ())
-}
-
-fn clear_raft_db(engine: &RocksEngine) -> Result<(), EngineError> {
-    let start_key = keys::raft_log_key(0, 0);
-    let end_key = keys::raft_log_key(u64::MAX, u64::MAX);
-    engine.compact_range("default", Some(&start_key), Some(&end_key), false, 2)?;
-
-    let range = Range::new(&start_key, &end_key);
-    engine.delete_ranges_cf("default", DeleteStrategy::DeleteFiles, &[range])?;
-    engine.sync()?;
-
-    let mut count = 0;
-    engine.scan(&start_key, &end_key, false, |_, _| {
-        count += 1;
-        Ok(true)
-    })?;
-    assert_eq!(count, 0);
-    Ok(())
-}
-
-/// Check the potential original raftdb directory and try to dump data out.
-///
-/// Procedure:
-///     1. Check whether the dump has been completed. If there is a dirty dir,
-///        delete the original raftdb safely and return.
-///     2. Scan and dump raft data into raft engine.
-///     3. Rename original raftdb dir to indicate that dump operation is done.
-///     4. Delete the original raftdb safely.
-pub fn check_and_dump_raft_db(
-    config: &TiKvConfig,
-    engine: &RaftLogEngine,
-    env: &Arc<Env>,
-    thread_num: usize,
-) {
-    let raftdb_path = &config.raft_store.raftdb_path;
-    let dirty_raftdb_path = get_path_for_remove(raftdb_path);
-    if !RocksEngine::exists(raftdb_path) {
-        remove_tmp_dir(&dirty_raftdb_path);
-        return;
-    }
-
-    // Clean the target engine if it exists.
-    clear_raft_engine(engine).expect("clear_raft_engine");
-
-    let config_raftdb = &config.raftdb;
-    let mut raft_db_opts = config_raftdb.build_opt();
-    raft_db_opts.set_env(env.clone());
-    let raft_db_cf_opts = config_raftdb.build_cf_opts(&None);
-    let db = engine_rocks::raw_util::new_engine_opt(raftdb_path, raft_db_opts, raft_db_cf_opts)
-        .unwrap_or_else(|s| fatal!("failed to create origin raft db: {}", s));
-    let src_engine = RocksEngine::from_db(Arc::new(db));
+pub fn dump_raftdb_to_raft_engine(source: &RocksEngine, target: &RaftLogEngine, threads: usize) {
```

Comment:
can we move these functions to `RaftDataStateMachine`? so the logic of the struct is more coherent.

Reply:
It's tempting. But I also think state machine should avoid depending on specific engine implementation.